### ISSUE TITLE
Update to Micronaut 2.0.0 breaks injection of DefaultSecurityService in background tasks

### ISSUE THREAD

**Issue by oscarfh** [micronaut-projects/micronaut-core#3868]
Hi,
I am migrating a project from Micronaut 1.x to 2.0 and I am facing a behavior that was changed and I am wondering whether this is something I can configure and that I should change entirely.

My code is something like this (simplified version)

```
public class MyClass() {
	@Inject
	private DefaultSecurityService defaultSecurityService;

	public CompletableFuture<Boolean> check() {
		Optional<Authentication> accessToken = defaultSecurityService.getAuthentication(); //Here is works
		return CompletableFuture.supplyAsync(() -> {
			Optional<Authentication> accessToken2 = defaultSecurityService.getAuthentication(); //Here is is Optional.empty
			// do some stuff, including calling a HttpClient
		}
	}
}
```

The problem is that I have a filter down the chain that uses `defaultSecurityService.getAuthentication();`. In 1.x, I was able to get the user's authentication and create a token for it. Now I can't and my request is made without token and fails.

My filter
```
@Filter("/path/**")
public class AuthHeaderFilter implements HttpClientFilter {
	@Inject
	private DefaultSecurityService defaultSecurityService;

	@Override
	public Publisher<? extends HttpResponse<?>> doFilter(MutableHttpRequest<?> request, ClientFilterChain chain) {
		return chain.proceed(request.bearerAuth(getToken(defaultSecurityService)); //I cannot generate the token here anymore
	}
```

Question:
**Can I propagate the user context to the new thread, so my filter works as before?**
(I need the security service in my filter, I added it to the check method in order to verify whether the behavior was due to the async thread or some change on the filter).

Clarifications:
* I am using Async because I make multiple http call async and then combine the result of the calls

**Steps to Reproduce:**
Checkout [this commit in this repository](https://github.com/oscarfh/micronaut-thread-context-inheritance/tree/22a46a77e8c3b58fbf65ab50f1e5c934b2eb0c2e) and Run the com.example.DemoTest#testHello test. You will be able to see to following output in your logs:
```
defaultSecurityService.getAuthentication(): Optional[io.micronaut.security.authentication.AuthenticationUserDetailsAdapter@772bace8]
defaultSecurityService.getAuthentication(): Optional[io.micronaut.security.authentication.AuthenticationUserDetailsAdapter@772bace8]
```
Which indicates that the context was inherited by the backgroup thread.

Then check out [this commit](https://github.com/oscarfh/micronaut-thread-context-inheritance/tree/3b8c4b114e1f1615887bc0857c153546e43970a5), **clean** and **rebuild ** and you will see the following output:
```
defaultSecurityService.getAuthentication(): Optional[io.micronaut.security.authentication.AuthenticationUserDetailsAdapter@2b471c18]
defaultSecurityService.getAuthentication(): Optional.empty
```
Indicating that the context is not propagated anymore.

**Comment by someshrathi02** [comment]
What is error you are getting..? if its circular dependency related then pls see below .. https://github.com/micronaut-projects/micronaut-core/issues/3239 
 You may need to use Provider<DefaultSecurityService>

**Comment by oscarfh** [comment]
@someshrathi02 No error, but `defaultSecurityService.getAuthentication()` returns `Optional.empty` instead of the authentication object.

**Comment by oscarfh** [comment]
@someshrathi02  I edited the original post with a project to reproduce the issue.

**Comment by someshrathi02** [comment]
Thanks foe updates. so its different issue than one which i faced while upgrading micronaut version

**Comment by dstepanov** [comment]
@oscarfh Can you use RxJava2's instead of `CompletableFuture::supplyAsync`?

**Comment by oscarfh** [comment]
@dstepanov yes, I did and it worked. Nevertheless, in my opinion, the fact that the context was passed to the new thread and now it is not anymore is an important behavior change that should, at least, be documented.

**Comment by graemerocher** [comment]
we probably need a way to handle this use case although I believe `CompletableFuture::supplyAsync` uses the fork join common pool which is not ideal for the use case described and is supposed to be used for short lived, non-blocking operations.

We may be able to support the use case whereby `supplyAsync` is used with a specific `ExecutorService` that you can wrap in instrumentation. Thoughts @dstepanov ?

**Comment by dstepanov** [comment]
Actually changing the example to
```
    @Inject
    @Named(TaskExecutors.IO)
    private ExecutorService executorService;
```
It correctly injects an instrumented Executor and context is preserved.

@graemerocher Looking at this commit https://github.com/micronaut-projects/micronaut-core/commit/928a2ba152a2d2419e218a9cd7bd632addb71c23 should it be `if (beanType == ExecutorService.class || beanType == ScheduledExecutorService.class)`?

**Comment by oscarfh** [comment]
@dstepanov @graemerocher If you feel it is more appropriate, feel free to close. 
Thanks a lot for the thread!

**Comment by ojebari-ma** [comment]
Can you please confirm that this is still an issue in the latest version of Micronaut?

**Comment by graemerocher** [comment]
Triaged as Bug for classification, then closed as not planned. The original behavior comes from using `CompletableFuture.supplyAsync` on the unmanaged ForkJoin common pool; Micronaut cannot automatically propagate request/security context into arbitrary executor tasks. Current Core propagates context through Micronaut-managed/instrumented executors and reactive/coroutine paths, so the supported fix is to inject a named `ExecutorService` (for example the IO executor), or explicitly propagate the required context. Reopen with a current 5.x example using a Micronaut-managed executor if context is still lost.
